[Feature] Improve Error Handling

kaimast opened the pull request:
## Motivation
Improve error handling in `snarkVM` to
* clearly differentiate user and I/O errors from internal errors that are bugs
* format `anyhow` error chains to show as much information as possible
* make testing and debugging easier (this PR serves a basis for #2789)
* provide helper functions for better error handling in snarkOS (see [snarkOS #3826](https://github.com/ProvableHQ/snarkOS/pull/3826))

## Changes
### Top-level panic handling
This PR now install a panic handler at `vm/cli/main.rs` and shows a "pretty" output when encountering panics. See the example below.
```
$> snarkvm execute foo

⚠️ snarkVM encountered an unexpected error at vm/cli/commands/execute.rs:49:20:
attempt to divide by zero

This is most likely a bug!
Please report it to the snarkVM developers: https://github.com/ProvableHQ/snarkVM/issues/new?template=bug.md
```

### Top-level handling for user errors
`snarkvm` will now print the entire chain of anyhow errors. For example, when passing invalid arguments to `snarkvm run` the following output is generated.

```
$> snarkvm run main 0u32 0u16
⚠️ Running leo_test.aleo:main with inputs [0u32, 0u16] failed
     ↳ 'u32' is invalid: expected u32, found 0u16

Use `--help` for instructions on how to use this command
```

## Other changes
* A `ensure_equals` macro that compares two values and `bail`s they do not match. This is used in the block validation functions to print the actual values when something goes wrong.
* Improvements to `ledger_test_helpers`
* A utility function to log an `anyhow::Error`. It "flattens" the error by joining the messages with an em dash. This is similar to how snarkOS currently handles such errors
* Avoid printing an empty line when there are no updates for snarkVM
* Minor fixes to the README

## Notes
* REST/Query changes were moved to #2901.
* As @ljedrz pointed out, we cannot use `panic = "abort"`, which was the case in an earlier version of the PR, due to how program errors are handled in snarkVM. Instead, the CLI uses a nested try_catch.
* This change just adds a panic handler as another layer of checks, and does not change existing errors in the code to panics. We would still need to discuss whether to allow panics in production code.

kaimast committed:
refactor: return anyhow error on invalid block

kaimast committed:
refactor: better anyhow Error handling

ljedrz commented on ledger/block/src/genesis.rs:
```diff
@@ -20,25 +20,32 @@ impl<N: Network> Block<N> {
-            && self.transactions.num_finalize() == 2 * Self::NUM_GENESIS_TRANSACTIONS
-            // Ensure there are no aborted transaction IDs in the genesis block.
-            && self.aborted_transaction_ids.is_empty()
+    pub fn is_genesis(&self) -> Result<bool> {
```
this is a very welcome change for development - I remember having related issues in the past and not knowing which bit went wrong :+1:

kaimast replied:
Yeah, this was the main reason I started this set of changes.

ljedrz reviewed (commented):
I like the general direction; what I (and probably the other reviewers) would like to see is some examples for how the common errors are displayed (e.g. the VM halting), and to know whether you've checked that the `locktick` feature (which relies on backtraces) still works with the `panic = "abort"` change introduced here.

kaimast committed:
refactor(console): use into_io_error to preserve error chain

kaimast committed:
refactor(algorithms): use into_io_error to preserve error chain

ljedrz commented on ledger/block/src/header/metadata/genesis.rs:
```diff
@@ -15,58 +15,67 @@
-            && self.last_coinbase_timestamp == N::GENESIS_TIMESTAMP
-            // Ensure the timestamp in the genesis block is `GENESIS_TIMESTAMP`.
-            && self.timestamp == N::GENESIS_TIMESTAMP
+    /// Returns `true` if the metadata is a genesis metadata.i
```
```suggestion
    /// Returns `true` if the metadata is a genesis metadata.
```

ljedrz commented on ledger/block/src/header/metadata/genesis.rs:
```diff
@@ -15,58 +15,67 @@
+            "Invalid cumulative proof target"
+        );
+        ensure_equals!(self.last_coinbase_timestamp, N::GENESIS_TIMESTAMP, "Invalid last coinbase timestamp");
+        ensure_equals!(self.last_coinbase_timestamp, N::GENESIS_TIMESTAMP, "Invalid last coinbase timestamp");
```
```suggestion
```
duplicate; probably meant to be `self.timestamp`?

vicsn replied:
This was marked as resolved but it still seems to be duplicate

kaimast replied:
I might have forgotten to push a commit. It is resolved now.

ljedrz commented on ledger/block/src/header/metadata/verify.rs:
```diff
@@ -33,78 +35,50 @@ impl<N: Network> Metadata<N> {
+            bail!("Metadata is malformed in block {expected_height}: {err}");
+        }
+
+        ensure_equals!(self.round, expected_round, "Round is incorrect in block {expected_height}");
```
this should be below the height check if we want to use the `expected_height` as authoritative, no?

kaimast replied:
I am unsure. I just set the checks in the order the arguments are passed to the function.

kaimast replied:
(It is also in the order that existed before this change)

vicsn replied:
Yeah I guess neither is authoritative and the height is just shared in the error message as context. I think the refactor looks good as is.

ljedrz commented on ledger/query/LICENSE.md:
is this intentional? I'd personally prefer to not have the copies of the license file everywhere, but it's the current convention

kaimast replied:
This change was moved to the other PR, but I would really prefer if we started to point to a single LICENSE file. This file was also formatted differently than the one in the root directory.

ljedrz reviewed (changes requested):
Could we split out the REST changes into a follow-up PR? While they build on the new abstractions from this one, they introduce some additional changes to that setup.

kaimast commented:
> Could we split out the REST changes into a follow-up PR? While they build on the new abstractions from this one, they introduce some additional changes to that setup.

Good point. I removed the commit for the REST changes from this PR and opened another draft PR for it (#2901).

kaimast committed:
chore: run cargo update

kaimast committed:
ci: fix compilation by changing cache key suffix

ljedrz reviewed (approved):
LGTM

vicsn commented on ledger/block/src/header/metadata/mod.rs:
```diff
@@ -80,40 +81,50 @@ impl<N: Network> Metadata<N> {
-                    && self.last_coinbase_timestamp >= N::GENESIS_TIMESTAMP
-                    // Ensure the timestamp in the block is after the genesis timestamp.
-                    && self.timestamp > N::GENESIS_TIMESTAMP
+    pub fn check_validity(&self) -> Result<()> {
```
Why rename this function? Please keep refactors to a minimum, especially given that you keep the other `is_valid`'s intact

kaimast replied:
The function signature is different. It returns `Result<()>` now, carrying more information on why the block is invalid.
I added `is_valid()` which simply calls `check_validity().is_ok()` for backwards compatibility.

vicsn commented on ledger/block/src/header/metadata/mod.rs:
```diff
@@ -80,40 +81,50 @@ impl<N: Network> Metadata<N> {
+        ensure!(self.last_coinbase_timestamp >= N::GENESIS_TIMESTAMP, "Ensure last coinbase timestamp");
+        ensure!(self.timestamp > N::GENESIS_TIMESTAMP, "Invalid timeestamp");
+
+        if self.coinbase_target < N::GENESIS_COINBASE_TARGET {
+            bail!(
+                "Invalid coinbase target: Was {actual} but expected {min} or greater.",
+                actual = self.coinbase_target,
+                min = N::GENESIS_COINBASE_TARGET,
+            );
+        }
+
+        if self.proof_target < N::GENESIS_PROOF_TARGET {
+            bail!(
+                "Invalid proof target: Was {actual} but expected {min} or greater.",
+                actual = self.proof_target,
+                min = N::GENESIS_PROOF_TARGET,
+            );
```
Why not stick to the ensure! and >/>= convention? Makes it more consistent and easier to read.

kaimast replied:
I made this change so I can see the actual values, and not just `invalid coinbase target` as the error message.

vicsn commented on ledger/block/src/header/mod.rs:
```diff
@@ -72,29 +74,27 @@ impl<N: Network> Header<N> {
-                    && self.ratifications_root != Field::zero()
-                    // Ensure the metadata is valid.
-                    && self.metadata.is_valid()
+    pub fn check_validity(&self) -> Result<()> {
```
Same, why go through the effort of changing the name? Mind you its also a pub function (like too many functions in snarkVM), so it has a chance of breaking some external utility.

kaimast replied:
Same answer as to the other one. I kept `is_valid` for backwards compatibility.

vicsn commented on ledger/src/advance.rs:
```diff
@@ -71,7 +74,8 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
                 rng,
-            )?;
+            )
+            .with_context(|| "Failed to construct block template")?;
```
```suggestion
            .with_context(|| "Failed to construct block")?;
```

kaimast replied:
This is message is generated when `construct_block_template` fails. I think it is correct?

vicsn commented on synthesizer/program/src/logic/instruction/operation/hash.rs:
```diff
@@ -239,7 +239,7 @@ macro_rules! do_hash {
 ///
 /// This allows running the hash without the machinery of stacks and registers.
-/// This is necessary for the Leo interpeter.
+/// This is necessary for the Leo interpreter.
```
This is not worth anyone's time to fix and review.

kaimast replied:
Sorry, my editor shows typos as warnings, so I correct it. I will not fix unrelated typos in future PRs.

vicsn replied:
Ah thx for sharing, that explains :)

vicsn commented on utilities/src/vm_error.rs:
```diff
@@ -23,22 +23,24 @@ macro_rules! try_vm_runtime {
             #[cfg(debug_assertions)]
-            eprintln!("{msg}");
+            {
+                let msg = e.to_string().replacen("panicked", "VM safely halted", 1);
```
Can you explain why you removed these lines:
```
           let msg = msg.split_ascii_whitespace().skip_while(|&word| word != "panicked").collect::<Vec<&str>>();
            let mut msg = msg.join(" ");
```

kaimast replied:
Must have been by mistake. I re-added the trimming.

vicsn commented on utilities/src/vm_error.rs:
```diff
@@ -23,22 +23,24 @@ macro_rules! try_vm_runtime {
-            eprintln!("{msg}");
+            {
+                let msg = e.to_string().replacen("panicked", "VM safely halted", 1);
+                eprintln!("{msg}");
```
Shall we wrap it in `dev-eprint`? 

Deferring a discussion of also printing trace logs to here: https://github.com/ProvableHQ/snarkVM/issues/2849

kaimast replied:
I think this might also be a good thing to tackle in a separate PR?

vicsn commented on vm/cli/main.rs:
```diff
@@ -13,19 +13,53 @@
-fn main() -> anyhow::Result<()> {
+use std::panic::catch_unwind;
+
+fn main() {
```
Shall we remove the entire snarkVM CLI in a next PR? CC @d0cd 

It is a chore to keep updated. In the meantime:
- `execute` is already available in snarkOS or any other app importing snarkVM
- `new` creates an aleo instruction program to deploy, which is not useful by itself and not something we care about exposing to users. For testing, unit tests suffice until we have a clear need in place.

This can be a separate issue.

kaimast replied:
Sounds good to me, but I would prefer doing this in a separate PR.

kaimast committed:
fix(ledger/block): add is_valid() for backwards compatiblity

kaimast committed:
fix(ledger/block): address reviewer comments

kaimast committed:
Merge remote-tracking branch 'origin/staging' into feat/better-errors

kaimast committed:
fix(utils): re-add trimming of panic message to try_vm_runtime
